### PR TITLE
Define rules for additional operations needed for HF model

### PR DESCRIPTION
**Pull request by migeed-z**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80976
* #80925
* #80909
* #80847
* #80823
* #80799
* #80744
* __->__ #80147

- Edit the constraint interface to also handle a pair `trace.root` and `graph` instead of a graphmodule
- Define all the operations needed for an example HFmodel
- Generate constraints for some of these operations. Those which are still in progress are marked with a todo comment. 

The operations we cover in this PR are:
- embedding
- size 
- view
- mul
- assert
- getitem

Differential Revision: [D37387596](https://our.internmc.facebook.com/intern/diff/D37387596)

### CODE REVIEW

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37387596).

File: torch/fx/experimental/migrate_gradual_types/constraint_generator.py [hunk 1]
```diff
@@ -33,31 +35,244 @@ def generate_flatten_constraints(start_dim, end_dim, input, flattened, n, counte
...
 
-@register_inference_rule(torch.add)
-@register_inference_rule(operator.add)
-def add_inference_rule(n: Node, symbols, constraints, counter):
+# TODO
+@register_inference_rule("long")
+def long_inference_rule(n: Node, symbols, constraints, counter):
+    """
+    """
+    return [], counter
```
Comment (jansel) [inline torch/fx/experimental/migrate_gradual_types/constraint_generator.py:43]: What does `[]` imply here? Why are there so many rules following this same empty pattern with no docstring?

A comment explaining how to read these rules (at the top) would be helpful.

If you are going to have a long list of rules with the same body, perhaps auto-generate them in a loop.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/fx/experimental/migrate_gradual_types/constraint_generator.py [hunk 2]
```diff
@@ -33,31 +35,244 @@ def generate_flatten_constraints(start_dim, end_dim, input, flattened, n, counte
...
+    """
+    assert isinstance(n.args[0], Node)
+
+    embedding_dim = module_instance.embedding_dim  # number
+
+    embedding_output, counter = gen_tvar(counter)
+    symbols[n] = embedding_output
+    embedding_input = symbols[n.args[0]]
+
+    input_dyn = BinConstraintT(embedding_input, Dyn, op_eq)
```
[the remark this answers is not in the document]
Comment (migeed-z) [reply torch/fx/experimental/migrate_gradual_types/constraint_generator.py:112]: Yes it does. The reason is that we want to generate constraints that cover all the possibilities for what they type can be. One possibility is that if the input to `Embedding` is` Dyn` then the output must also be `Dyn`. We combine those two expressions in a constraint ` c1 = Conj([input_dyn, output_dyn])` then we consider the possibility of the input being a tensor, which is in `c2` where we consider all possibilities of the input being a tensor and consider what the output would be in each case. For this function, assuming a tensor input, the output shape differs from the input shape in the last dimension. So we encode that for every case.

 In the end we combine everything into one constraint which is `[Disj([c1, Disj(c2)])]`.

File: torch/fx/experimental/migrate_gradual_types/constraint_generator.py [hunk 3]
```diff
@@ -33,31 +35,244 @@ def generate_flatten_constraints(start_dim, end_dim, input, flattened, n, counte
...
+    get_item_output, counter = gen_dvar(counter)
+    symbols[n] = get_item_output
+
+    # retreive arg variables
+    get_item_arg = symbols[n.args[0]]
+
+    # if the input is dynamic, we accept any index and return
+    # a dynamic dimension as output
+    input_dyn = BinConstraintT(get_item_arg, Dyn, op_eq)
+    output_dyn = BinConstraintD(get_item_output, Dyn, op_eq)
```
[the remark this answers is not in the document]
Comment (migeed-z) [reply torch/fx/experimental/migrate_gradual_types/constraint_generator.py:198]: D will be translated to a variable that Z3 will interpret as a number, for dimensions. 
T will be translated to a variable that Z3 will interpret as a tensor. So when we generate a set of constraints, Z3 can find a satisfying assignment according to those variations. We have them since certain constraints are specifically over tensors and others are over dimensions, where we need to do arithmetic over them.

File: test/fx/test_z3_gradual_types.py [hunk 4]
```diff
@@ -28,6 +28,56 @@
...
+
+        B = BasicBlock().forward(torch.ones([2, 4], dtype=torch.long)).size()
+        ast_rewriter = RewritingTracer()
+        graph = ast_rewriter.trace(BasicBlock())
+        traced = GraphModule(ast_rewriter.root, graph, "gm")
+
+        transformed = transform_all_constraints(traced, counter=0)
+        s = z3.Solver()
+        s.add(transformed)
+        assert s.check() == z3.sat
```
[the remark this answers is not in the document]
Comment (migeed-z) [reply test/fx/test_z3_gradual_types.py]: done

File: test/fx/test_z3_gradual_types.py [hunk 5]
```diff
@@ -28,6 +28,77 @@
...
```
[the remark this answers is not in the document]
Comment (migeed-z) [reply test/fx/test_z3_gradual_types.py]: done

File: torch/fx/experimental/migrate_gradual_types/constraint_generator.py (hunk 1 above)
Comment (migeed-z) [reply torch/fx/experimental/migrate_gradual_types/constraint_generator.py:43]: These are the list of rules we still need to do. I left them empty and I plan to write them in the next PRs. `[]` implies no constraint.

File: torch/fx/experimental/migrate_gradual_types/constraint_generator.py [hunk 6]
```diff
@@ -33,31 +35,244 @@ def generate_flatten_constraints(start_dim, end_dim, input, flattened, n, counte
...
+
+
+@register_inference_rule(torch.nn.modules.sparse.Embedding)
+def embedding_inference_rule(n: Node, module_instance, symbols, constraints, counter):
+    """
+    The output shape differs from the input shape in the last dimension
+    """
+    assert isinstance(n.args[0], Node)
+
+    embedding_dim = module_instance.embedding_dim  # number
```
[the remark this answers is not in the document]
Comment (migeed-z) [reply torch/fx/experimental/migrate_gradual_types/constraint_generator.py:106]: which dim? is it something other than the embedding dim?

For example this gives an error:
`torch.nn.Embedding(256008, -2, padding_idx=1)
`

File: torch/fx/experimental/migrate_gradual_types/constraint_generator.py [hunk 7]
```diff
@@ -33,31 +35,244 @@ def generate_flatten_constraints(start_dim, end_dim, input, flattened, n, counte
...
-@register_inference_rule(operator.add)
-def add_inference_rule(n: Node, symbols, constraints, counter):
+# TODO
+@register_inference_rule("long")
+def long_inference_rule(n: Node, symbols, constraints, counter):
+    """
+    """
+    return [], counter
+
+# TODO
```
Comment (jamesr66a) [inline torch/fx/experimental/migrate_gradual_types/constraint_generator.py:45]: Maybe `throw NotImplementedError` in the body of these TODO functions?

Comment (migeed-z) [reply torch/fx/experimental/migrate_gradual_types/constraint_generator.py:45]: done

**Review by jamesr66a — APPROVED** [review]
